### PR TITLE
SALTO-1748: add restrictions to field type overrides

### PR DESCRIPTION
**Pull request by noamhammer**
_Add the option to set restrictions on specific fields on duck type config_

---
---
_Release Notes_: 
_Add the options to set restrictions on fields via the config on duck type adapters_

---

### CODE REVIEW

File: packages/adapter-components/src/elements/field_type_config_override.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { FieldTypeOverrideType } from '../config/transformation'
+import { getTypeTransformationConfig, TypeConfig, TypeDefaultsConfig } from '../config/shared'
+
+const log = logger(module)
+
+export const fixFieldTypes = (
+  definedTypes: Record<string, ObjectType>,
+  typeConfig: Record<string, TypeConfig>,
+  typeDefaultConfig: TypeDefaultsConfig,
+  fieldTypeNameToType?: (typeName: string) => TypeElement,
```
Comment (netama) [inline packages/adapter-components/src/elements/field_type_config_override.ts:27]: why do we need to pass in `toTypeWithContainers` as optional? what's the risk of having it work the same way for ducktype? we probably didn't need it until now, but AFAICT the implementation should work for it as well, no?

File: packages/adapter-components/src/elements/field_type_config_override.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        return
+      }
+      const fieldTypeOverrides = getTypeTransformationConfig(
+        typeName, typeConfig, typeDefaultConfig,
+      ).fieldTypeOverrides as FieldTypeOverrideType[]
+      fieldTypeOverrides.forEach(({ fieldName, fieldType, restrictions }) => {
+        const field = type.fields[fieldName]
+        const newFieldType = fieldTypeNameToType
+          ? fieldTypeNameToType(fieldType)
+          : (definedTypes[fieldType] || primitiveTypes[fieldType])
```
Comment (netama) [inline packages/adapter-components/src/elements/field_type_config_override.ts:50]: same question as above - is there a reason to limit this and not allow list/map types?

File: packages/adapter-components/src/elements/swagger/type_elements/type_config_override.ts [hunk 3]
```diff
@@ -113,44 +114,18 @@ export const fixTypes = (
...
-      })
-    })
+  fixFieldTypes(definedTypes, typeConfig, typeDefaultConfig, toTypeWithContainers)
 
   // hide env-specific fields
   Object.entries(definedTypes)
-    .filter(([typeName]) => getTypeTransformationConfig(typeName).fieldsToHide !== undefined)
+    .filter(([typeName]) =>
+      getTypeTransformationConfig(typeName, typeConfig, typeDefaultConfig)
+        .fieldsToHide !== undefined)
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/type_elements/type_config_override.ts:124]: readability nit:
```suggestion
      getTypeTransformationConfig(
        typeName, typeConfig, typeDefaultConfig,
      ).fieldsToHide !== undefined)
```
(and same below)

File: packages/adapter-components/test/elements/ducktype/transformer.test.ts [hunk 4]
```diff
@@ -381,7 +381,14 @@ describe('ducktype_transformer', () => {
...
 
-    beforeEach(() => {
+    afterEach(() => {
+      jest.clearAllMocks()
+    })
+    afterAll(() => {
+      jest.restoreAllMocks()
+    })
+
+    it('should return the type, nested types and instances', async () => {
```
Comment (netama) [inline packages/adapter-components/test/elements/ducktype/transformer.test.ts:391]: is this intentional?

File: packages/adapter-api/src/builtins.ts [hunk 5]
```diff
@@ -112,7 +112,7 @@ const dependencyType = new ObjectType({
   },
 })
 
-type RestrictionAnnotationType = Partial<{
+export type RestrictionAnnotationType = Partial<{
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/adapter-api/src/builtins.ts:115]: should be ok to export, it is already in the language spec so it was kind of in the interface even without being exported

File: packages/adapter-components/test/elements/ducktype/transformer.test.ts (hunk 4 above)
Comment (noamhammer) [reply packages/adapter-components/test/elements/ducktype/transformer.test.ts:391]: yes, moved the spyOn from the beforeEach to the function itself

File: packages/adapter-components/src/elements/field_type_config_override.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        const annotations = restrictions ? {
```
[the remark this answers is not in the document]
Comment (noamhammer) [reply packages/adapter-components/src/elements/field_type_config_override.ts]: @alonstern and I talked about it and we decided to override it. So keep it like that

**Review by netama — APPROVED** [review]

**Pull request merged**